Issue 8: Vault and Strategy Holdings

miguel567 opened the pull request:
Issue: https://github.com/iearn-finance/yearn-api/issues/8

- Added ./abi folder as a volume to docker-compose.yml file to ease development.
- calculated Vault holdings and Strategy holdings and Pools Balance (virtualPrice * Totalsupply)
- Storing price_usd from coingecko for vault's want tokens.
- Created a new table called holdings.

I'm following the approach from dudesahn as defined here: hackmd.io/@dudesahn/BkxKfTzqw
TODO: We believe we are missing to calculate the pool balance of yWBTC token.
This is the pre-requirement for calculating TVL (my next step).

miguel567 commented:
I'm breaking APY/SAVE. need to fix it first.

miguel567 commented:
fixed the APY save endpoint which was broken before and removed ethereum-blocks-by-date.js file as this is not needed.

x48-crypto commented on services/vaults/apy/save/handler.js:
```diff
@@ -276,7 +276,7 @@ const readVault = async (vault) => {
 };
 
-module.exports.handler = async () => {
+const handler = async () => {
```
why change this? was it broken?

miguel567 replied:
At the end of the file I have set:
module.exports = { handler, getVirtualPrice };

The reason is that I'm reusing the same getVirtualPrice function in : https://github.com/miguel567/yearn-api/blob/91c1696f5f5fa309126883fd17a51ee964a1448b/services/vaults/holdings/save/getHoldings.js#L94

so I had to modify the exports module to include both functions.

x48-crypto commented on README.md:
```diff
@@ -32,37 +32,6 @@ git clone https://github.com/yearn-integrations/yearn-api.git
-
-## Stages
-
-- Currently four stages are available
```
what is intention behind removing this README section?

miguel567 replied:
I totally believe this is a mistake. Shoud I fix it and push again?

miguel567 replied:
solved

x48-crypto commented on services/vaults/holdings/save/handler.js:
```diff
@@ -0,0 +1,111 @@
+  }
+  try{
+    const holdings= await getHoldings(vault);
+    const priceFeed = await axios.get('https://api.coingecko.com/api/v3/coins/' + vault.price_id);
```
can we get price on-chain from uniswap instead perhaps? this way we don't have to maintain the API

miguel567 replied:
OK, I need to investigate how to perform that. Any guidance?

miguel567 replied:
Can we do this on an improvement ticket? Unless it is a blocker, I could make the other improvements easily, deliver the TVL endpoint and then focus on changing the price feed.

miguel567 replied:
the "earn" products: yWBTC, yCRV and ybCRV are using getVirtualPrice = on-chain price.
Rest of tokens are using coingecko. I would merge as it is and create a new issue to move all pricing infro to on-chain using chainlink, for example. issue: https://github.com/iearn-finance/yearn-api/issues/34

dudesahn replied:
We should be good, we can just use Andre's keep3r price oracles, which already have all of the pairs we need. I updated with the relevant info on issue #34

x48-crypto commented on services/vaults/holdings/save/vaults.js:
```diff
@@ -0,0 +1,293 @@
+
+
+module.exports = [
+  {
```
Can we get vaults from vaults endpoint instead? The vaults endpoint pulls from the vault registry

miguel567 replied:
ah yes, I saw you improved that part, I think I can do this easily.

miguel567 replied:
Icannot use vaults from DB as the ABI is not stored. I could use:
`
const fetchAbi = async (address) => {
  const url = `https://api.etherscan.io/api?module=contract&action=getabi&address=${address}&apikey=${etherscanApiKey}`;
  const resp = await fetch(url).then((res) => res.json());
  const abi = JSON.parse(resp.result);
  await delay(delayTime);
  return abi;
};
`
from services/vaults/holdings/save/handler.js, but we would rely on etherscan API.

miguel567 replied:
As discussed in discord this cannot be done for now. We cannot fetch the ABI on-chain. Also, the services/vaults/holdings/save/vaults.js file is defining where the ABI is coming from. Also I cannot remove the vault.js file as, for instance the aLink and Link strategies have 6 decimals and there is no on-chain "decimals" method for these strategies for me to fetch the correct decimals. So for now we need to rely on the code in the files and not on-chain data :S

x48-crypto commented on services/vaults/save/handler.js:
```diff
@@ -1,14 +1,14 @@
 const yRegistryAbi = require('../../../abi/yRegistry.json');
 const delay = require('delay');
-
+const db = dynamodb.doc;
```
nice fix, i spotted this as well

miguel567 committed:
Merge pull request #1 from iearn-finance/master

update

graham-u commented on services/vaults/holdings/save/getHoldings.js:
```diff
@@ -0,0 +1,119 @@
+  const vaultContract = new web3.eth.Contract(vault.vaultContractABI, vault.vaultContractAddress);
+  const vaultHoldings = await vaultContract.methods.balance().call() / Math.pow(1,vault.decimals);
+  const strategyHoldings = await strategyContract.methods.balanceOf().call() / Math.pow(1,vault.decimals);
+    /* console.log(holdings); */
```
Can we remove commented out console.log?

miguel567 replied:
Updated in the new push all the commented out code.

graham-u commented on services/vaults/holdings/save/getHoldings.js:
```diff
@@ -0,0 +1,119 @@
+
+
+    return holdings;  
+}
```
Need an empty line at end of file.

miguel567 replied:
done.

graham-u commented on services/vaults/holdings/save/handler.js:
```diff
@@ -0,0 +1,111 @@
+    .catch((err) => console.log("err", err));
+  console.log(`Saved ${data.name}`);
+};
+
+
+
+
```
Personal opinion but I think 4 blank lines here is a lot, better to have 1 or 2.  Same above as well.

miguel567 replied:
removed

graham-u reviewed (commented):
There are quite a few instances of commented out code that I think we should remove unless it is there for a good reason.

miguel567 commented:
I've reviewed your feedback @graham-u  and @x48-crypto  form the previous push, please check it above.

Also I've added to this push:
- YFI staked holdings calculations
- veCRV holdings calculations.
- Added the compound Vault holdings calculations (I will add it to APY endpoint in this issue: https://github.com/iearn-finance/yearn-api/issues/32)


FYI: I'm following Dudesahn's documented holdings calculation guide: https://hackmd.io/@dudesahn/BkxKfTzqw

miguel567 committed:
feat: implement vault holdings endpoints
